### PR TITLE
909 adherence status control

### PR DESCRIPTION
**Pull request by clemens-msupply**
To test:
- go to earliest encounter and set "Quantity prescribed" to 7 (Quantity prescribed is actually pills on hand after the current encounter, better name?)
- save
- got to next encounter and modify pills remaining

Also includes:
- some API clean up
- Numeric min value input fixes

### CODE REVIEW

**Comment by clemens-msupply** [comment]
There are some encounter query improvements in https://github.com/openmsupply/open-msupply/pull/934....
I.e. it now uses cached encounter queries

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    return null;
+  }
+
+  const inputProps = {
+    InputProps: {
+      sx: { width: '90px', '& .MuiInput-input': { textAlign: 'right' } },
+    },
+    disabled: true,
+    error: !!errors,
+    helperText: errors,
```
Comment (CarlosNZ) [inline client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx:149]: Was thinking we should probably also show a specific error state for illegal (negative) values. If the user enters a higher "Remaining Pill Count" than there are available, then the adherence value is negative. It's probably too much hassle to have an additional check on the "Remaing pill count" field in order to prevent illegal values (would have to do a lookup to previous encounter to define its maximum), so might be easier just to show something informative (in red) in this component instead. (perhaps "N/A"?)

File: client/packages/programs/src/api/api.ts [hunk 2]
```diff
@@ -34,10 +43,39 @@ export const getEncounterQueries = (sdk: Sdk, storeId: string) => ({
...
+    patientId: string,
+    current: Date
+  ): Promise<EncounterBaseFragment> => {
+    const result = await sdk.encounters({
+      storeId,
+      key: EncounterSortFieldInput.StartDatetime,
+      desc: true,
+      filter: {
+        startDatetime: {
+          beforeOrEqualTo: new Date(current.getTime() - 1).toISOString(),
```
Comment (CarlosNZ) [inline client/packages/programs/src/api/api.ts:69]: Something for later -- might want to add a strictly `before` filter field so we don't have to worry about the `-1`. 
Not a priority though. :)

**Review by CarlosNZ — APPROVED** [review]
Great, works nicely and quite easy to understand the code. :) 

Have added a few small comments and questions, but most probably don't require changes on your part.

One other thought -- don't know if we want to address this here or think about it for later. This component is only looking at the previous encounter's "Quantity prescribed" field. This means that on the *third* (and subsequent) encounter, unless the practitioner prescribed *more* pills, the continuing adherence rate won't be calculated. But it seems like something that should continue to work regardless of whether more were prescribed (especially since we have the graph showing adherence rate over time). (And if the patient has not been taking their pills, then they're less likely to be prescribed more)

Perhaps the "remainingCount" in this component could also consider the previous encounter's "remaining pills" entry as well as the "quantity prescribed". (Add them together?) I can see potential problems with this, so maybe we should discuss this when @richardmoizeau  gets back? (Unless you can think of another more obvious immediate solution?)

File: server/service/src/sync/program_schemas/hiv_care_encounter_ui_schema.json [hunk 3]
```diff
@@ -108,6 +108,30 @@
...
           "label": "Remaining Pill Count"
         },
+        {
+          "scope": "#",
+          "type": "AdherenceStatus",
+          "label": "Adherence Status",
+          "options": {
+            "previousCountField": "arvMedication.quantityPrescribed",
+            "remainingCountField": "arvMedication.remainingPillCount",
+            "countPerDay": 1,
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/sync/program_schemas/hiv_care_encounter_ui_schema.json:118]: This is a constant, e.g. you have to take 2 pills per day. In our case its one pill a day

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+
+  useEffect(() => {
+    if (!options) {
+      return;
+    }
+    const targetPath = composePaths(path, options.targetField);
+    setTargetPath(targetPath);
+    setAdherenceStatus(extractProperty(data, targetPath));
+  }, [options, path]);
+  useEffect(() => {}, [targetPath]);
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx]: thanks this is an error fixed in: [d515c68](https://github.com/openmsupply/open-msupply/pull/921/commits/d515c68912be8c63f82308329fcadce98c4c90b5)

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+  // fetch current encounter
+  const encounterId = useEncounter.utils.idFromUrl();
+  const { data: currentEncounter, mutate: fetchEncounter } =
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx:72]: Good catch. The mutate is used to fetch the encounter right now without caching . I used it because this was how the hook was already implemented. In a stacked PR https://github.com/openmsupply/open-msupply/pull/934 I already changed it to use `useQuery` so that the encounter isn't queried multiple times to render the same view...

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+      currentEncounter?.patient.id ?? '',
+      currentEncounter?.startDatetime
+        ? new Date(currentEncounter?.startDatetime)
+        : new Date()
+    );
+  useEffect(() => {
+    if (currentEncounter) {
+      fetchPreviousEncounter();
+    }
+  }, [currentEncounter]);
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx:90]: Agree was a bit copy & paste without thinking... As mentioned the next PR has a better solution using `react-query`

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx [hunk 7]
```diff
@@ -0,0 +1,172 @@
...
+
+  const params:
+    | {
+        previousCountOnHand: number;
+        remainingCount: number;
+        countPerDay: number;
+        timeDiffMs: number;
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx:97]: agree 'Ms" looks slightly odd but still prefer to have the unit in the var name especially since I then transform it to days...

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+      previousEncounter.document.data,
+      options.previousCountField
+    );
+    const remainingCount = extractProperty(data, options.remainingCountField);
+    const timeDiffMs =
+      new Date(currentEncounter.startDatetime).getTime() -
+      new Date(previousEncounter.startDatetime).getTime();
+
+    // Target pill count needed for the whole timespan from last till current encounter
+    const timeDiffDays = timeDiffMs / 1000 / 60 / 60 / 24;
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx:114]: isn't it harder to keep the comment in sync with the code, e.g. when fixing a conversion mistake?

Think we should define some HOUR, DAY constants somewhere...

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx (hunk 1 above)
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx:148]: Yes good point, not sure about it, maybe some warning? lets see if it could be a real issue I will check in meeting today

File: client/packages/programs/src/api/api.ts (hunk 2 above)
Comment (clemens-msupply) [reply client/packages/programs/src/api/api.ts:69]: yes! a `before` and `after` filter is missing

**Comment by clemens-msupply** [comment]
> Great, works nicely and quite easy to understand the code. :)
> 
> Have added a few small comments and questions, but most probably don't require changes on your part.
> 
> One other thought -- don't know if we want to address this here or think about it for later. This component is only looking at the previous encounter's "Quantity prescribed" field. This means that on the _third_ (and subsequent) encounter, unless the practitioner prescribed _more_ pills, the continuing adherence rate won't be calculated. But it seems like something that should continue to work regardless of whether more were prescribed (especially since we have the graph showing adherence rate over time). (And if the patient has not been taking their pills, then they're less likely to be prescribed more)
> 
> Perhaps the "remainingCount" in this component could also consider the previous encounter's "remaining pills" entry as well as the "quantity prescribed". (Add them together?) I can see potential problems with this, so maybe we should discuss this when @richardmoizeau gets back? (Unless you can think of another more obvious immediate solution?)

Another good point and I have the same problem for ARV status changes where you might not change the status in each encounter. Think I have a solution, lets discuss in person maybe tomorrow...

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx [hunk 9]
```diff
@@ -15,6 +16,7 @@ export const PositiveNumberInput = React.forwardRef(
       sx,
       disabled = false,
       value,
+      min = 1,
       max = 999999999,
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx]: I used `Number.MAX_SAFE_INTEGER` and also ensure that the input stays positive...

@mark-prins could you please have look as well at this PositiveNumberInput change? (min option has been added to set min to 0)

[c12efb3](https://github.com/openmsupply/open-msupply/pull/921/commits/c12efb3b1bc2dc93cfd68355565ae8f3ad967bdb)

File: server/service/src/sync/program_schemas/hiv_care_encounter_ui_schema.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/service/src/sync/program_schemas/hiv_care_encounter_ui_schema.json:118]: Okay, so this "Medication" section is for a specific drug/dosage in this Program type?

[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/sync/program_schemas/hiv_care_encounter_ui_schema.json:118]: yes

Just checked the doc comment for this option in TypeScript:
"Expected number of pills per day that a patient is suppose to take."

do you think this is clear enough?

[the remark this answers is not in the document]
Comment (CarlosNZ) [reply server/service/src/sync/program_schemas/hiv_care_encounter_ui_schema.json:118]: Yeah, the message is clear enough. It's not clear that its a fixed value though. I assumed that different patients would receive different dosages, but I don't really know what I'm talking about, medicine-wise.

[the remark this answers is not in the document]
Comment (clemens-msupply) [reply server/service/src/sync/program_schemas/hiv_care_encounter_ui_schema.json:118]: yes for this specific case it the same dose for all patients

File: client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/programs/src/JsonForms/components/AdherenceStatus.tsx:114]: Nice idea. maybe an enum in `DateUtils`?

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (mark-prins) [reply client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx]: The reason for the lower value was that we had `Number.MAX_SAFE_INTEGER` originally and that caused errors in the API due to  the rust maximum integer value being lower, which meant that the supplied number couldn't be parsed as u32. [JS](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) `2^53 - 1` and [Rust](https://doc.rust-lang.org/std/primitive.u32.html#associatedconstant.MAX): `2^32 - 1`

We can't know what the API is trying to parse the number as, so safest to use a sensible minimum. Perhaps add a CONST value in common.

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx [hunk 10]
```diff
@@ -15,7 +16,8 @@ export const PositiveNumberInput = React.forwardRef(
       sx,
       disabled = false,
       value,
-      max = 999999999,
+      min = 1,
+      max = Number.MAX_SAFE_INTEGER,
```
Comment (mark-prins) [inline client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx:20]: Can you revert this change please? It was the source of a bug previously, as noted.

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx [hunk 11]
```diff
@@ -28,7 +30,9 @@ export const PositiveNumberInput = React.forwardRef(
         InputProps={{
           sx: { ...sx, '& .MuiInput-input': { textAlign: 'right' } },
         }}
-        onChange={value => onChange(NumUtils.constrain(value, 1, max))}
+        onChange={value =>
+          onChange(NumUtils.constrain(value, Math.max(0, min), max))
```
Comment (mark-prins) [inline client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx:34]: Can you revert this one too please - the usage of this input elsewhere expects a 1 minimum e.g. packsize which should never be zero. 

Perhaps add a parameter to set the minimum value?

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx]: cool thanks will do, half expected some reason behind it...

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx (hunk 10 above)
Comment (clemens-msupply) [reply client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx:20]: thanks done in:
https://github.com/openmsupply/open-msupply/pull/945

File: client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx (hunk 11 above)
Comment (clemens-msupply) [reply client/packages/common/src/ui/components/inputs/TextInput/numeric/PositiveNumberInput.tsx:34]: There is already a min parameter that defaults to 1 so it should behave like before (otherwise its a bug)
